### PR TITLE
Feature/log filtering

### PR DESCRIPTION
**Pull request by jjelschen**

### CODE REVIEW

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/PersistenceModule.xtend [hunk 1]
```diff
@@ -10,6 +10,8 @@ import org.testeditor.web.backend.testexecution.loglines.LogFinder
 import org.testeditor.web.backend.testexecution.loglines.ScanningLogFinder
 import org.testeditor.web.backend.testexecution.screenshots.ScreenshotFinder
 import org.testeditor.web.backend.testexecution.screenshots.SubStepAggregatingScreenshotFinder
+import org.testeditor.web.backend.testexecution.loglines.LogFilter
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/persistence/PersistenceModule.xtend]: organize imports :-)

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilter.xtend [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.regex.Pattern
+
+import static org.testeditor.web.backend.testexecution.loglines.LogLevel.*
+
+class Log4JDefaultFilter implements LogFilter {
+
+	static val log4jLevelNames = #{'FATAL' -> CRITICAL, 'ERROR' -> ERROR, 'WARN' -> WARNING, 'INFO' -> INFO, 'DEBUG' -> DEBUG, 'TRACE' -> TRACE}
+	static val log4jPattern = Pattern.compile(
+		'\\s+\\d{1,2}:\\d{1,2}:\\d{1,2}\\s+([A-Z]{4,5})\\s+\\[Test worker\\]\\s+\\[TE-Test:\\s+[^\\]]+\\]\\s+[^\\s]+.*')
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilter.xtend]: We might want the pattern to be less specific, maybe? Testing on "Test worker" looks a bit too specific to me.

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilter.xtend [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+	override isVisibleOn(String logLine, LogLevel logLevel) {
+		if (logLevel === TRACE) {
+			return true
+		} else {
+			val matcher = log4jPattern.matcher(logLine)
+			return matcher.find && log4jLevelNames.get(matcher.group(1)) <= logLevel
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilter.xtend]: is this compare valid? the matcher returns a String, doesn't it?

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilterTest.xtend [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		INFO -> '    23:42:00 INFO  [Test worker]  [TE-Test: TestName] ReportingClass Message part.',
+		DEBUG -> '    23:42:00 DEBUG [Test worker]  [TE-Test: TestName] ReportingClass Message part.',
+		TRACE -> '    23:42:00 TRACE [Test worker]  [TE-Test: TestName] ReportingClass Message part.'
+	}
+	static val unformattedLogLine = 'FATAL ERROR WARN INFO DEBUG TRACE unformatted log line'
+
+	@Parameters(name='log line: "{0}"; log level: {1}; accept? {2}')
+	def static Collection<Object[]> data() {
+		return #[
+			#[logLines.get(CRITICAL), CRITICAL, true],
```
Comment (gunther-bachmann) [inline org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilterTest.xtend:35]: I would expect the result of isVisibleOn to return exactly the opposite actually. If I ask for CRITICAL I would expect only critical messages, if I ask for Errors, I'd expect CRITICAL and ERROR, asking for INFO, I'd expect CRITICAL, ERROR, WARNING and INFO (but no DEBUG and no TRACE).

**Review by gunther-bachmann — CHANGES_REQUESTED** [review]

File: org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilter.xtend (hunk 3 above)
Comment (jjelschen) [reply org.testeditor.web.backend.persistence/src/main/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilter.xtend]: Yeah, but I am using `log4jLevelNames` to map it to an enum value, so its a comparison between compatible types (`LogLevel`). The rest is Xtend magic, which in turn maps the `<=` to a call to `compareTo(…)` – here's the corresponding, generated Java code:
```      return (matcher.find() && (Log4JDefaultFilter.log4jLevelNames.get(matcher.group(1)).compareTo(logLevel) <= 0));```

… I did miss the case that the matcher finds a match, but the matched group does not contain a valid log level. In this case, I now assume a default log level of TRACE.

File: org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilterTest.xtend (hunk 4 above)
Comment (jjelschen) [reply org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilterTest.xtend:35]: I think that is what the method does… isn't it?
Notice that the first argument is the actual log entry, and the second is the desired log level. This allows to write the method "extension style" and make it more readable (or so I thought :)
```
'23:42:00 WARN [Test worker] …'.isVisibleOn(INFO) // true, WARN is more severe than INFO
'23:42:00 WARN [Test worker] …'.isVisibleOn(WARN) // true
'23:42:00 WARN [Test worker] …'.isVisibleOn(ERROR) // false, WARN is less severe than ERROR
```

Comment (gunther-bachmann) [reply org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/testexecution/loglines/Log4JDefaultFilterTest.xtend:35]: Ahh, I finally got it

**Review by gunther-bachmann — APPROVED** [review]

**Pull request merged**